# **Book Recommendation System**

## **About DataSet**

The Book-Crossing dataset comprises 3 files.

Users
- Contains the users. User IDs. Demographic data is provided (Location, Age) if available. Otherwise, these fields contain NULL-values.

Books
- Books are identified by their respective ISBN. Moreover, some content-based information is given (Book-Title, Book-Author, Year-Of-Publication, Publisher), obtained from Amazon Web Services. Note that in case of several authors, only the first is provided. URLs linking to cover images are also given, appearing in three different flavours (Image-URL-S, Image-URL-M, Image-URL-L), i.e., small, medium, large. These URLs point to the Amazon web site.

Ratings
- Contains the book rating information. Ratings (Book-Rating) are either explicit, expressed on a scale from 1-10 (higher values denoting higher appreciation), or implicit, expressed by 0.


DataSet Link:-
http://www2.informatik.uni-freiburg.de/~cziegler/BX/

## **Importing Necessary Libraries**

In [28]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings("ignore")

# **Loading Data File**

In [29]:
books = pd.read_csv('/content/drive/MyDrive/Portfolio/Book_Recomender/BX-Books.csv', sep=';', error_bad_lines=False, encoding='latin-1')

users = pd.read_csv('/content/drive/MyDrive/Portfolio/Book_Recomender/BX-Users.csv', sep=';', error_bad_lines=False, encoding='latin-1')

ratings = pd.read_csv('/content/drive/MyDrive/Portfolio/Book_Recomender/BX-Book-Ratings.csv', sep=';', error_bad_lines=False, encoding='latin-1')

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'


## **Understanding the data**

### How does the data look like ?

In [30]:
books.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [31]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [32]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

### How big is the data ?

In [33]:
print("Shape of books :", books.shape)
print("Shape of users :  ", users.shape)
print("Shape of ratings :  ", ratings.shape) 

Shape of books : (271360, 8)
Shape of users :   (278858, 3)
Shape of ratings :   (1149780, 3)


### Are there any missing values ?

In [34]:
books.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [35]:
users.isna().mean()*100

User-ID      0.000000
Location     0.000000
Age         39.719857
dtype: float64

In [36]:
ratings.isna().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

### Are there any duplicate values ?

In [37]:
books.duplicated().sum()

0

In [38]:
users.duplicated().sum()

0

In [39]:
ratings.duplicated().sum()

0

## **Insight**

**Dataset consists of 3 different data files : 'Books, Users, Ratings'.**

**Shape of dataset is :**
- Books : (271360, 8)
- Users :   (278858, 3)
- Ratings :   (1149780, 3)

**DataSet has missing values.**

**There are 39% missing values in 'age' feature in 'Users' dataset.**
- We will drop the feature as age is not required to build our recommender system and it also consists more missing values.

# **EDA**

## **Books**

In [40]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [41]:
# Selecting the required features for the model

books = books[['ISBN','Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher']]
books.sample(5)

ISBN                                         Book-Title  \
11806   014070728X            King Henry IV (New Penguin Shakespeare)   
90977   0704311658                                               Dare   
125007  0071378227  Darwin In the Genome: Molecular Strategies in ...   
62920   0345358120                                     More Puzzicles   
16808   0880702753  Kids With Character: Preparing Children for a ...   

                 Book-Author Year-Of-Publication  \
11806    William Shakespeare                1981   
90977     Philip Jose Farmer                1965   
125007        Lynn  Caporale                2002   
62920           Alan Robbins                1989   
16808   Marti Watson Garlett                1989   

                             Publisher  
11806                    Penguin Books  
90977                    Quartet Books  
125007  McGraw-Hill/Contemporary Books  
62920                 Ballantine Books  
16808                    Multnomah Pub

In [42]:
# Renaming the features for better understanding

books.rename(columns={'Book-Title':'title', 'Book-Author':'author', 'Year-Of-Publication':'year', 'Publisher':'publisher'}, inplace=True)
books.head()

ISBN                                              title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

                 author  year                   publisher  
0    Mark P. O. Morford  2002     Oxford University Press  
1  Richard Bruce Wright  2001       HarperFlamingo Canada  
2          Carlo D'Este  1991             HarperPerennial  
3      Gina Bari Kolata  1999        Farrar Straus Giroux  
4       E. J. W. Barber  1999  W. W. Norton &amp; Company

## **Users**

In [43]:
# Selecting the required features for the model
users = users[['User-ID','Location']]
users.sample(5)

User-ID                             Location
180240   180241            hamburg, hamburg, germany
245600   245601                   dallas, texas, usa
264551   264552              brumath, alsace, france
200708   200709         middletown, connecticut, usa
57616     57617  perth, western australia, australia

In [44]:
# Renaming the features for better understanding

users.rename(columns={'User-ID':'user_id', 'Location':'location'}, inplace=True)
users.head()

user_id                            location
0        1                  nyc, new york, usa
1        2           stockton, california, usa
2        3     moscow, yukon territory, russia
3        4           porto, v.n.gaia, portugal
4        5  farnborough, hants, united kingdom

## **Ratings**

In [45]:
# Renaming the features for better understanding

ratings.rename(columns={'User-ID':'user_id', 'Book-Rating':'rating'}, inplace=True)
ratings.head()

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5
2   276727  0446520802       0
3   276729  052165615X       3
4   276729  0521795028       6

### **Rechecking shape of the dataset**

In [46]:
print("Shape of books :", books.shape)
print("Shape of users :  ", users.shape)
print("Shape of ratings :  ", ratings.shape)

Shape of books : (271360, 5)
Shape of users :   (278858, 2)
Shape of ratings :   (1149780, 3)


## **Treating missing values**

In [47]:
books.isna().sum()

ISBN         0
title        0
author       1
year         0
publisher    2
dtype: int64

In [48]:
# Dropping missing values by CCA method
# As we cannot fill the 'author' and 'publisher' of the book
# There no point in filling the data by 'Mode' or by writting 'Missing' inplace

books.dropna(axis=0, inplace=True)

In [49]:
books.isna().sum()

ISBN         0
title        0
author       0
year         0
publisher    0
dtype: int64

# **Visualisation**

In [50]:
#!pip install pandas-profiling 
#!pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip

from pandas_profiling import ProfileReport

## **Books**

In [51]:
# year col is unsupported

In [52]:
prof1 = ProfileReport(books)

prof1.to_file(output_file = 'output1.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## **Users**

In [53]:
prof2 = ProfileReport(users)

prof2.to_file(output_file = 'output2.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## **Ratings**

# **Preprocessing**

We will consider those users who have rated >=200 books
- to get knowledable and genuine users.

We will only consider books which have been rated by >=50 users
- to get good quality books.


## **Collaborative Filtering Technique**

- If there is user : 'A' and user : 'B' and if they both have read a book - '1'
- But user : 'A' also have read book - '2'
- So we will recommend the book - '2' to user : 'B'

**We will use Pivot_table**

- features/col: users_id
- indices/rows: title
- values : ratings

In [54]:
ratings['user_id'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: user_id, Length: 105283, dtype: int64

In [55]:
ratings['user_id'].nunique()

105283

### **Cleaning the data**

In [56]:
# Considering those users who have rated >=200 books

rating = ratings['user_id'].value_counts()>200
rating[rating].shape

(899,)

---------------------------------------------------------------------------
**There are only 899 users who have rated more than 200 books.**

---------------------------------------------------------------------------

In [57]:
# Getting the index of required users

y = rating[rating].index
y

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            260183,  73681,  44296, 155916,   9856, 274808,  28634,  59727,
            268622, 188951],
           dtype='int64', length=899)

In [58]:
# Filtering the users in ratings feature

new_ratings = ratings[ratings['user_id'].isin(y)]
new_ratings.head()

user_id        ISBN  rating
1456   277427  002542730X      10
1457   277427  0026217457       0
1458   277427  003008685X       8
1459   277427  0030615321       0
1460   277427  0060002050       0

In [59]:
new_ratings.shape

(526356, 3)

### **Merging new_ratings with books**

In [60]:
# Merging datasets on 'ISBN' feature

ratings_books = new_ratings.merge(books)
ratings_books.head()

user_id        ISBN  rating  \
0   277427  002542730X      10   
1     3363  002542730X       0   
2    11676  002542730X       6   
3    12538  002542730X      10   
4    13552  002542730X       0   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
2  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
3  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
4  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   publisher  
0  John Wiley &amp; Sons Inc  
1  John Wiley &amp; Sons Inc  
2  John Wiley &amp; Sons Inc  
3  John Wiley &amp; Sons Inc  
4  John Wiley &amp; Sons Inc

In [61]:
ratings_books.shape

(487669, 7)

In [62]:
# Considering the books with >50 ratings

ratings_books.groupby('title')['rating'].count()

title
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    2
 Always Have Popsicles                                                                                        1
 Apple Magic (The Collector's series)                                                                         1
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   1
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                1
                                                                                                             ..
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                           3
Ã?Â?lpiraten.                                                                                                 1
Ã?Â?rger mit Produkt X. Roman.                                                                    

In [63]:
number_rating = ratings_books.groupby('title')['rating'].count().reset_index()

In [64]:
number_rating.rename(columns={'rating':'number_of_ratings'}, inplace=True)
number_rating.head()

title  number_of_ratings
0   A Light in the Storm: The Civil War Diary of ...                  2
1                              Always Have Popsicles                  1
2               Apple Magic (The Collector's series)                  1
3   Beyond IBM: Leadership Marketing and Finance ...                  1
4   Clifford Visita El Hospital (Clifford El Gran...                  1

### **Merging ratings_books with number_rating**

In [65]:
# Merging datasets on 'title' feature

final_rating = ratings_books.merge(number_rating)

In [66]:
final_rating.shape

(487669, 8)

In [67]:
# Finding out the books with >50 raings

final_rating = final_rating[final_rating['number_of_ratings']>=50]
final_rating.sample(5)

user_id        ISBN  rating  \
116922   228998  0671867091       0   
5162     263877  0451180410       0   
89137     30533  0380717018       8   
6367     232131  0439064864       9   
76155    147965  0671673688       0   

                                                    title  \
116922                                        Remember Me   
5162                                          Mortal Fear   
89137   Body of Evidence (Kay Scarpetta Mysteries (Pap...   
6367     Harry Potter and the Chamber of Secrets (Book 2)   
76155                          While My Pretty One Sleeps   

                      author  year    publisher  number_of_ratings  
116922    Mary Higgins Clark  1995       Pocket                 87  
5162               Greg Iles  1998  Signet Book                 61  
89137   Patricia D. Cornwell  1992         Avon                 69  
6367           J. K. Rowling  1999   Scholastic                183  
76155     Mary Higgins Clark  1990       Pocket                 75

In [68]:
final_rating.shape

(61853, 8)

### **Droping duplicate rows if any**

In [69]:
final_rating.drop_duplicates(['user_id', 'title'], inplace=True)
final_rating.shape

(59850, 8)

## **Making pivot_table**

In [70]:
book_pivot = final_rating.pivot_table(columns='user_id', index='title', values='rating')
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [71]:
# Checking shape of pivot_table data

book_pivot.shape

(742, 888)

---------------------------------------------------------------------------
- **We have total 742 books**
- **Total users are 888**

--------

Our 'users' are filtered as those users were the users, who have rated the books which were not rated by more than 50 users.
- Shape of users chsnged from 899 to 888 now.

# **Filling Nan values**

In [72]:
# Filling Nan values with 0
# 0(zero) indicates that user has not rated that book

book_pivot.fillna(0, inplace=True)

## **Final Dataset**

In [73]:
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

# **Modeling**

In [74]:
# Using 'Sparse - Matrix'; so that distances with 0 should not be considered/calculated

from scipy.sparse import csr_matrix

In [75]:
# Creating 'Sparse - Matrix',
# This will also reduce the computation power

book_sparse = csr_matrix(book_pivot)
book_sparse

<742x888 sparse matrix of type '<class 'numpy.float64'>'
	with 14942 stored elements in Compressed Sparse Row format>

In [76]:
type(book_sparse)

scipy.sparse.csr.csr_matrix

## **Clustering**

### **Using KNN**

In [77]:
# Importing KNN Algorithm
from sklearn.neighbors import NearestNeighbors


# Using algo = 'BRUTE', will use a brute-force search to calculate all distances
# n_neighbours = 5 (default)
model = NearestNeighbors(algorithm='brute')

In [78]:
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

### **Defining the recommendation**

In [79]:
book_pivot.index[237]

'Harry Potter and the Chamber of Secrets (Book 2)'

In [80]:
# input :- book value
# no_of_recommendations = n+1 (considers the distance)

distances, suggestions=model.kneighbors(book_pivot.iloc[237,:].values.reshape(1,-1), n_neighbors=6)

In [81]:
# provides the distance with other books in asc
distances

array([[ 0.        , 68.78953409, 69.5413546 , 72.64296249, 76.83098333,
        77.28518616]])

In [82]:
# provides the index of nearest distances
suggestions

array([[237, 240, 238, 241, 184, 536]])

In [83]:
book_pivot.index[237]

'Harry Potter and the Chamber of Secrets (Book 2)'

### **Output of Recommended Books**

In [84]:
for i in range(len(suggestions)):
  print(book_pivot.index[suggestions[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')


In [85]:
book_pivot.index[240]

'Harry Potter and the Prisoner of Azkaban (Book 3)'

In [86]:
distances, suggestions=model.kneighbors(book_pivot.iloc[240,:].values.reshape(1,-1), n_neighbors=6)

In [87]:
for i in range(len(suggestions)):
  print(book_pivot.index[suggestions[i]])

Index(['Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Sorcerer's Stone (Book 1)',
       'Harry Potter and the Order of the Phoenix (Book 5)', 'Exclusive'],
      dtype='object', name='title')


In [88]:
book_pivot.index[54]

'Animal Farm'

In [89]:
distances, suggestions=model.kneighbors(book_pivot.iloc[54,:].values.reshape(1,-1), n_neighbors=6)

In [90]:
for i in range(len(suggestions)):
  print(book_pivot.index[suggestions[i][1:]])

Index(['Exclusive', 'Jacob Have I Loved', 'Second Nature', 'Pleading Guilty',
       'No Safe Place'],
      dtype='object', name='title')


# **Assigning Function for Recommendation**

In [91]:
np.where(book_pivot.index=='Animal Farm')         # it's 2-dim array

(array([54]),)

In [92]:
def recommend_book(book_name):
  book_id = np.where(book_pivot.index==book_name)[0][0]       # [0][0] to extract index of book
  distances, suggestions = model.kneighbors(book_pivot.iloc[book_id, :].values.reshape(1,-1), n_neighbors=6)

  for i in range(len(suggestions)):
    if i==0:
      print("The suggestions for ", book_name, "are : ")
    if not i:
      print(book_pivot.index[suggestions[i][1:]])

In [93]:
recommend_book('Big Trouble')

The suggestions for  Big Trouble are : 
Index(['No Safe Place', 'Long After Midnight', 'Exclusive',
       'Lake Wobegon days', 'Journey'],
      dtype='object', name='title')
